nrclient/proxy.go

```diff
@@ -0,0 +1,214 @@
+		transport.DialTLS = fallbackDialer(transport)
+	}
+
+	return transport, nil
+}
+
+func getCertPool(certFile string, certDirectory string) (*x509.CertPool, error) {
+	caCertPool := systemCertPool()
+
+	if certFile != "" {
+		caCert, err := ioutil.ReadFile(certFile)
+		if err != nil {
+			return nil, fmt.Errorf("can't read certificate file %v: %v", certFile, err)
+		}
+
+		ok := caCertPool.AppendCertsFromPEM(caCert)
+		if !ok {
+			log.Printf("[WARN] certificates from %v could not be appended", certFile)
+		}
+	}
+	if certDirectory != "" {
+		files, err := ioutil.ReadDir(certDirectory)
+		if err != nil {
+			return nil, fmt.Errorf("can't read certificate directory %v: %v", certDirectory, err)
+		}
+
+		for _, f := range files {
+			if strings.Contains(f.Name(), ".pem") {
+				caCertFilePath := filepath.Join(certDirectory + "/" + f.Name())
+				caCert, err := ioutil.ReadFile(caCertFilePath)
+				if err != nil {
+					return nil, fmt.Errorf("can't read certificate file %v: %v", caCertFilePath, err)
+				}
+				ok := caCertPool.AppendCertsFromPEM(caCert)
+				if !ok {
+					log.Printf("[WARN] certificate %v could not be appended", caCertFilePath)
+				}
+			}
+		}
+	}
+	return caCertPool, nil
+}
+
+func getProxyResolver(ignoreSystemProxy bool, proxy string) (func(*http.Request) (*url.URL, error), error) {
+	if len(proxy) > 0 {
+		// User-defined nrclient
+		prUrl, err := url.Parse(proxy)
+		if err != nil {
+			return nil, err
+		}
+
+		return http.ProxyURL(prUrl), nil
+	} else if !ignoreSystemProxy {
+		// Proxy defined via the HTTPS_PROXY (takes precedence) or HTTP_PROXY environment variables
+		return http.ProxyFromEnvironment, nil
+	} else {
+		// No nrclient
+		return http.ProxyURL(nil), nil
+	}
+}
+
+func resolveProxyURL (proxyResolver func(*http.Request) (*url.URL, error), nrEndpoint string) (*url.URL, error){
+	nrEndpointURL, err := url.Parse(nrEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("can't parse the NR logging endpoint, please contact New Relic: %v", err)
+	}
+	nrUrlRequest := http.Request{
+		URL: nrEndpointURL,
+	}
+
+	return proxyResolver(&nrUrlRequest)
+}
+
+// Dial verifier implements the transport.Dialer interface to provide backwards compatibility with Go 1.9 nrclient
+// implementation.
+//
+// It does the following process:
+//
+// 1. Tries to normally connect to an HTTPS nrclient
+// 2. If succeeds, uses the normal `tls.Dial` function in further connections
+// 3. If an Unknown Authority Error is returned, InsecureSkipVerify is set to true and we continue using
+//    `tls.Dial` for the following connections.
+// 4. If the secure connection is not accepted, we use an unsecured "Go1.9-like" dialer that does not
+//    performs the TLS handshake.
+//
+// IMPORTANT: This verification mode should be only done with legacy nrclient implementation, where the
+// proxy_validate_certificates configuration option is set to false, to avoid breaking changes with legacy users since the
+// update from Go 1.9 to Go 1.10.
```

Comment:
what is this comment trying to tell me? :)

Reply:
I tried to improve the comment, let me know what you think.